From e8d70afc7af94814655cfc213ac19201ac06c295 Mon Sep 17 00:00:00 2001
From: Rune Lillesveen <futhark@chromium.org>
Date: Thu, 07 Nov 2024 13:07:03 +0000
Subject: [PATCH] [@container] Make style() queries for IACVT values false

Also fix crash for querying IACVT value for registered property.

Bug: 376834363
Change-Id: Ief7718b633690992b3720b0100743f0f84c0525b
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5999463
Reviewed-by: Anders Hartvoll Ruud <andruud@chromium.org>
Commit-Queue: Rune Lillesveen <futhark@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1379633}
---

--- a/third_party/blink/renderer/core/css/properties/longhands/custom_property.cc
+++ b/third_party/blink/renderer/core/css/properties/longhands/custom_property.cc
@@ -97,6 +97,10 @@
 }
 
 void CustomProperty::ApplyInherit(StyleResolverState& state) const {
+  if (!state.ParentStyle()) {
+    ApplyInitial(state);
+    return;
+  }
   ComputedStyleBuilder& builder = state.StyleBuilder();
   bool is_inherited_property = IsInherited();
 
@@ -128,6 +132,7 @@
 
   if (value.IsInvalidVariableValue()) {
     if (!SupportsGuaranteedInvalid()) {
+      state.SetHasUnsupportedGuaranteedInvalid();
       ApplyUnset(state);
       return;
     }
@@ -189,6 +194,7 @@
   }
 
   if (!registered_value) {
+    state.SetHasUnsupportedGuaranteedInvalid();
     if (is_inherited_property) {
       ApplyInherit(state);
     } else {
--- a/third_party/blink/renderer/core/css/resolver/style_resolver.cc
+++ b/third_party/blink/renderer/core/css/resolver/style_resolver.cc
@@ -2818,6 +2818,9 @@
       set, {.origin = CascadeOrigin::kAuthor});
   cascade.Apply();
 
+  if (state.HasUnsupportedGuaranteedInvalid()) {
+    return nullptr;
+  }
   CSSPropertyRef property_ref(property_name, element->GetDocument());
   const ComputedStyle* style = state.TakeStyle();
   return ComputedStyleUtils::ComputedPropertyValue(property_ref.GetProperty(),
--- a/third_party/blink/renderer/core/css/resolver/style_resolver.h
+++ b/third_party/blink/renderer/core/css/resolver/style_resolver.h
@@ -208,6 +208,9 @@
 
   static bool CanReuseBaseComputedStyle(const StyleResolverState& state);
 
+  // Return a computed value for the passed-in property:value pair in the
+  // context of the current ComputedStyle of the 'element'.
+  // Returns nullptr for custom property values that are IACVT.
   static const CSSValue* ComputeValue(Element* element,
                                       const CSSPropertyName&,
                                       const CSSValue&);
--- a/third_party/blink/renderer/core/css/resolver/style_resolver_state.h
+++ b/third_party/blink/renderer/core/css/resolver/style_resolver_state.h
@@ -260,6 +260,13 @@
   void SetHasTreeScopedReference() { has_tree_scoped_reference_ = true; }
   bool HasTreeScopedReference() const { return has_tree_scoped_reference_; }
 
+  void SetHasUnsupportedGuaranteedInvalid() {
+    has_unsupported_guaranteed_invalid_ = true;
+  }
+  bool HasUnsupportedGuaranteedInvalid() const {
+    return has_unsupported_guaranteed_invalid_;
+  }
+
  private:
   CSSToLengthConversionData UnzoomedLengthConversionData(const FontSizeStyle&);
 
@@ -348,6 +355,10 @@
   // True if the resolved ComputedStyle depends on tree-scoped references.
   bool has_tree_scoped_reference_ = false;
 
+  // Tried to apply a guaranteed-invalid value to a custom property that doesn't
+  // support it.
+  bool has_unsupported_guaranteed_invalid_ = false;
+
   bool has_attr_function_ = false;
 };
 
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-conditional/container-queries/style-query-guaranteed-invalid.tentative.html
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
+<title>CSS Container Queries Test: style() query with guaranteed-invalid value</title>
+<link rel="help" href="https://drafts.csswg.org/css-conditional-5/#style-container">
+<link rel="help" href="https://github.com/w3c/csswg-drafts/issues/11143">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="support/cq-testcommon.js"></script>
+<style>
+  #target { background-color: green; }
+  @container --root style(--prop: var(--unknown)) {
+    #target { background-color: red; }
+  }
+</style>
+<div id="target"></div>
+<script>
+  setup(() => assert_implements_style_container_queries());
+
+  test(() => {
+    assert_equals(getComputedStyle(target).backgroundColor, "rgb(0, 128, 0)");
+  }, "Querying with a guaranteed-invalid value should evaluate to false");
+
+</script>
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-conditional/container-queries/style-query-registered-custom-invalid.tentative.html
@@ -0,0 +1,28 @@
+<!DOCTYPE html>
+<title>CSS Container Queries Test: style() query with IACVT value for registered custom property</title>
+<link rel="help" href="https://drafts.csswg.org/css-conditional-5/#style-container">
+<link rel="help" href="https://github.com/w3c/csswg-drafts/issues/11143">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="support/cq-testcommon.js"></script>
+<style>
+  :root { container-name: --root; }
+  @property --prop {
+    syntax: "a";
+    initial-value: a;
+    inherits: true;
+  }
+  #target { background-color: green; }
+  @container --root style(--prop: b) {
+    #target { background-color: red; }
+  }
+</style>
+<div id="target"></div>
+<script>
+  setup(() => assert_implements_style_container_queries());
+
+  test(() => {
+    assert_equals(getComputedStyle(target).backgroundColor, "rgb(0, 128, 0)");
+  }, "Querying IACVT value for registered custom property should evaluate to false");
+
+</script>
